### PR TITLE
feat(rulesets): add rules for validation of server variables and channel parameters

### LINKED ISSUE #1102: [AsyncAPI] Enhance parameters and variables linting
In #974, it's been identified that new rules could really be helpful. 

Especially:
- add a rule to ensure all variables used in server url are defined
- add a rule to ensure all defined server variables are used
- add a rule to ensure all parameters used in channel path are defined
- add a rule to ensure all defined channel parameters are used

The oas `path-params` custom function may be a great source of inspiration to fulfill that.

/cc @derberg

### PR DESCRIPTION
**Pull request by magicmatatjahu**
Fixes #1102

**Checklist**

- [X] Tests added / updated
- [X] Docs added / updated

**Does this PR introduce a breaking change?**

- [ ] Yes
- [X] No

**Context**

Two new rules add the ability to validate variables in server address and parameters in channel address. All variables/parameters must be defined and there can be no redundant variables/parameters.

> NOTE: workflows responsible for linting code and markdowns show errors, but in files which I have not changed at all... 🤷🏼

### CODE REVIEW

File: packages/rulesets/src/asyncapi/index.ts [hunk 1]
```diff
@@ -55,6 +57,17 @@ export default {
         },
       },
     },
+    'asyncapi-channel-parameters-defined': {
```
Comment (magicmatatjahu) [inline packages/rulesets/src/asyncapi/index.ts]: I don't know if it's a good name, or maybe simpler `asyncapi-channel-parameters` will be better.

File: packages/rulesets/src/asyncapi/index.ts [hunk 2]
```diff
@@ -282,6 +295,17 @@ export default {
         function: asyncApi2DocumentSchema,
       },
     },
+    'asyncapi-server-variables-defined': {
```
Comment (magicmatatjahu) [inline packages/rulesets/src/asyncapi/index.ts]: I don't know if it's a good name, or maybe simpler `asyncapi-server-variables` will be better.

File: packages/rulesets/src/asyncapi/index.ts (hunk 1 above)
Comment (marbemac) [reply packages/rulesets/src/asyncapi/index.ts]: 👍 IMHO `asyncapi-channel-parameters` is nicer (and more consistent with similar rules in other rulesets).

File: packages/rulesets/src/asyncapi/index.ts (hunk 2 above)
Comment (marbemac) [reply packages/rulesets/src/asyncapi/index.ts]: Ditto prior comment - obv a bit subjective, but for the same reasons I think `asyncapi-server-variables` is a better choice.

File: packages/rulesets/src/asyncapi/functions/utils.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
+export function parseUrlVariables(str: string): string[] | undefined {
```
Comment (marbemac) [inline packages/rulesets/src/asyncapi/functions/utils.ts]: Personal preference, but I appreciate how each top level file in the `functions` folder represents a single named function. Adding a `functions/utils.ts` breaks the convention - any chance we could move these to a nested `utils` folder, similar to how it's organized in `packages/rulesets/src/oas/functions`?

**Review by marbemac — COMMENTED** [review]
One small thought re code org, otherwise the implementation looks pretty good! @P0lip, any other thoughts?

Also @P0lip do you know what's up with the lint check?

File: packages/rulesets/src/asyncapi/index.ts (hunk 1 above)
Comment (magicmatatjahu) [reply packages/rulesets/src/asyncapi/index.ts]: Thanks! Changed :)

File: packages/rulesets/src/asyncapi/index.ts (hunk 2 above)
Comment (magicmatatjahu) [reply packages/rulesets/src/asyncapi/index.ts]: Thanks! Changed :)

File: packages/rulesets/src/asyncapi/functions/utils.ts (hunk 3 above)
Comment (magicmatatjahu) [reply packages/rulesets/src/asyncapi/functions/utils.ts]: Ok 👍🏼 I moved helper functions to the `asyncapi/functions/utils` alongside with unit tests. Thanks!

**Comment by magicmatatjahu** [comment]
@marbemac @P0lip Hi! Sorry for such delay but I had a lot of other things to do in our organization. I will try to deliver most of the rules that are missing by the end of next week. I propagated review suggestions. Thanks again for the review!

File: packages/rulesets/src/asyncapi/functions/asyncApi2ServerVariables.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+import { parseUrlVariables } from './utils/parseUrlVariables';
+import { getMissingProps } from './utils/getMissingProps';
+import { getRedundantProps } from './utils/getRedundantProps';
+
+import type { IFunctionResult } from '@stoplight/spectral-core';
+
+export default createRulesetFunction<{ url: string; variables: Record<string, unknown> }, null>(
+  {
+    input: null,
```
Comment (P0lip) [inline packages/rulesets/src/asyncapi/functions/asyncApi2ServerVariables.ts]:
```suggestion
    input: {
      type: 'object',
      properties: {
        url: {
          type: 'string',
        },
        variables: {
          type: 'object',
        }
      },
      required: ['url', 'variables'],
    },
```
let's validate the input, so the function is executed only when `url` and `variables` are the values we expect

File: packages/rulesets/src/asyncapi/functions/asyncApi2ChannelParameters.ts [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+import { parseUrlVariables } from './utils/parseUrlVariables';
+import { getMissingProps } from './utils/getMissingProps';
+import { getRedundantProps } from './utils/getRedundantProps';
+
+import type { IFunctionResult } from '@stoplight/spectral-core';
+
+export default createRulesetFunction<{ parameters: Record<string, unknown> }, null>(
+  {
+    input: null,
```
Comment (P0lip) [inline packages/rulesets/src/asyncapi/functions/asyncApi2ChannelParameters.ts]:
```suggestion
    input: {
      type: 'object',
      properties: {
        parameters: {
          type: 'object',
        }
      },
      required: ['parameters'],
    },
```

**Review by P0lip — COMMENTED** [review]
Left 2 comments. Nice work!

**Comment by magicmatatjahu** [comment]
@P0lip Sorry for such a delay! All suggestions have been applied. Let me know if everything is ok and if it is ok, you can merge when you want :) Unfortunately I had to do a mixed reset + force push (it was the easiest way for me) because the commit linting was showing me errors when I merged from the Github UI develop branch to my PR.

**Review by P0lip — APPROVED** [review]
Thanks for addressing the comments!

**Pull request merged**